File: src/douf/douf_multi_foreach.ss/tests/cpp/EntityHandler.cpp
/******************************************************************************
*
* Copyright Saab AB, 2008 (http://www.safirsdk.com)
* 
* Created by: Saab AB
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

#include "App.h"
#include "EntityHandler.h"
#include <Capabilities/Vehicles/Vehicle.h>
#include <Capabilities/Animals/Animal.h>
#include <Safir/Dob/SuccessResponse.h>
#include <Safir/Dob/ErrorResponse.h>
#include <Safir/Dob/EntityIdResponse.h>
#include <Safir/Dob/NotFoundException.h>
#include <Safir/Dob/ResponseGeneralErrorCodes.h>
#include <Safir/SwReports/SwReport.h>
#include <Safir/Time/TimeProvider.h>

#include <ace/OS_NS_unistd.h>
#include <boost/date_time/posix_time/posix_time_duration.hpp>

namespace VehicleAppCpp
{
    EntityHandler::EntityHandler() : m_debug(L"ForEachTest")
    {
        m_debug.Enable(true);
    }

    void EntityHandler::Init()
    {
        m_connection.Attach();

        // Register as vehicle entity handler.

        m_connection.RegisterEntityHandler(
            Capabilities::Vehicles::Vehicle::ClassTypeId, 
            Safir::Dob::Typesystem::HandlerId(),
            Safir::Dob::InstanceIdPolicy::RequestorDecidesInstanceId,
            this);

        m_connection.RegisterEntityHandler(
            Capabilities::Animals::Animal::ClassTypeId, 
            Safir::Dob::Typesystem::HandlerId(),
            Safir::Dob::InstanceIdPolicy::RequestorDecidesInstanceId,
            this);

         m_debug << "Registered entity handlers" << std::endl;
    }

    void EntityHandler::OnRevokedRegistration(
        const Safir::Dob::Typesystem::TypeId typeId,
        const Safir::Dob::Typesystem::HandlerId& handlerId)
    {
        // No longer registered for given type.
        Safir::SwReports::SendErrorReport(
            L"Unexpected revoked registration",
            L"EntityHandler::OnRevokedRegistration",
            L"The handler " + handlerId.ToString() +  L" is no longer registered for type " + 
            Safir::Dob::Typesystem::Operations::GetName(typeId));
    }

    void EntityHandler::OnCreateRequest(
        const Safir::Dob::EntityRequestProxy entityRequestProxy,
        Safir::Dob::ResponseSenderPtr responseSender)
    {
        bool bOk = false;
        Safir::Dob::Typesystem::InstanceId instanceId;
        Safir::Dob::Typesystem::EntityId entityId;

        const Capabilities::Vehicles::VehiclePtr vehicle =
            boost::dynamic_pointer_cast<Capabilities::Vehicles::Vehicle>(entityRequestProxy.GetRequest());

        if (vehicle != NULL)
        {

            //instanceId = Safir::Dob::Typesystem::InstanceId(Safir::Dob::Typesystem::InstanceId::GenerateRandom());

            // Check if entity with given value already exist.
            entityId = Safir::Dob::Typesystem::EntityId(
                Capabilities::Vehicles::Vehicle::ClassTypeId,
                entityRequestProxy.GetInstanceId());

            if(!m_connection.IsCreated(entityId))
            {
                // Store object in the Dob.
                m_connection.SetAll(vehicle, entityRequestProxy.GetInstanceId(), Safir::Dob::Typesystem::HandlerId());
                bOk = true;
            }

            if (bOk)
            {
                // Inform requestor about the instance.
               /* Safir::Dob::EntityIdResponsePtr entIdResponse = Safir::Dob::EntityIdResponse::Create();
                entIdResponse ->Assigned().SetVal(entityId);
                responseSender -> Send(entIdResponse);*/
                responseSender -> Send(Safir::Dob::SuccessResponse::Create());
            }
            else
            {
                Safir::Dob::ErrorResponsePtr errorResponse = Safir::Dob::ErrorResponse::Create();
                errorResponse -> Code().SetVal(Safir::Dob::ResponseGeneralErrorCodes::SafirReqErr());
                responseSender -> Send(errorResponse);
            }

        }

        const Capabilities::Animals::AnimalPtr animal =
            boost::dynamic_pointer_cast<Capabilities::Animals::Animal>(entityRequestProxy.GetRequest());

        if (animal != NULL)
        {
            //instanceId = Safir::Dob::Typesystem::InstanceId(Safir::Dob::Typesystem::InstanceId::GenerateRandom());

            // Check if entity with given value already exist.
            entityId = Safir::Dob::Typesystem::EntityId(
                Capabilities::Animals::Animal::ClassTypeId,
                entityRequestProxy.GetInstanceId());

            if(!m_connection.IsCreated(entityId))
            {
                // Store object in the Dob.
                m_connection.SetAll(animal, entityRequestProxy.GetInstanceId(), Safir::Dob::Typesystem::HandlerId());
                bOk = true;
            }


            if (bOk)
            {
                // Inform requestor about the instance.
                /*Safir::Dob::EntityIdResponsePtr entIdResponse = Safir::Dob::EntityIdResponse::Create();
                entIdResponse ->Assigned().SetVal(entityId);
                responseSender -> Send(entIdResponse);*/
                responseSender -> Send(Safir::Dob::SuccessResponse::Create());
            }
            else
            {
                Safir::Dob::ErrorResponsePtr errorResponse = Safir::Dob::ErrorResponse::Create();
                errorResponse -> Code().SetVal(Safir::Dob::ResponseGeneralErrorCodes::SafirReqErr());
                responseSender -> Send(errorResponse);
            }
        }
    }

    void EntityHandler::OnUpdateRequest(
        const Safir::Dob::EntityRequestProxy entityRequestProxy,
        Safir::Dob::ResponseSenderPtr responseSender)
    {
        bool bOk = false;
        
        m_debug << "OnUpdateRequest" << std::endl;
        boost::posix_time::ptime pt = Safir::Time::TimeProvider::ToLocalTime(Safir::Time::TimeProvider::GetUtcTime());
        m_debug << boost::posix_time::to_simple_wstring(pt) << std::endl;


        const Capabilities::Vehicles::VehiclePtr vehicle =
            boost::dynamic_pointer_cast<Capabilities::Vehicles::Vehicle>(entityRequestProxy.GetRequest());

        if (vehicle != NULL)
        {

            if (m_connection.IsCreated(entityRequestProxy.GetEntityId()))
            {
                // Update the stored vehicle with the received one.
                m_debug << "Sleeping for 1 seconds" << std::endl;
                ACE_OS::sleep(1);
                m_debug << "Done sleeping" << std::endl;
                // Update the stored vehicle with the received one.
                m_connection.SetChanges(
                    vehicle,
                    entityRequestProxy.GetInstanceId(),
                    Safir::Dob::Typesystem::HandlerId());
                bOk = true;
            }

            if (bOk)
            {
                responseSender -> Send(Safir::Dob::SuccessResponse::Create());
            }
            else
            {
                Safir::Dob::ErrorResponsePtr errorResponse = Safir::Dob::ErrorResponse::Create();
                errorResponse -> Code().SetVal(Safir::Dob::ResponseGeneralErrorCodes::SafirReqErr());
                responseSender -> Send(errorResponse);
            }
        }


        const Capabilities::Animals::AnimalPtr animal =
            boost::dynamic_pointer_cast<Capabilities::Animals::Animal>(entityRequestProxy.GetRequest());

        if (animal != NULL)
        {

            if(m_connection.IsCreated(entityRequestProxy.GetEntityId()))
            {
                // Update the stored vehicle with the received one.
                m_debug << "Sleeping for 5 seconds" << std::endl;
                ACE_OS::sleep(5);
                m_debug << "Done sleeping" << std::endl;
                // Update the stored vehicle with the received one.
                m_connection.SetChanges(
                    animal,
                    entityRequestProxy.GetInstanceId(),
                    Safir::Dob::Typesystem::HandlerId());
                bOk = true;
            }


            if (bOk)
            {
                responseSender -> Send(Safir::Dob::SuccessResponse::Create());
            }
            else
            {
                Safir::Dob::ErrorResponsePtr errorResponse = Safir::Dob::ErrorResponse::Create();
                errorResponse -> Code().SetVal(Safir::Dob::ResponseGeneralErrorCodes::SafirReqErr());
                responseSender -> Send(errorResponse);
            }
        }
    }


    void EntityHandler::OnDeleteRequest(
        const Safir::Dob::EntityRequestProxy entityRequestProxy,
        Safir::Dob::ResponseSenderPtr responseSender)
    {

        m_debug << "OnDeleteRequest" << std::endl;
        boost::posix_time::ptime pt = Safir::Time::TimeProvider::ToLocalTime(Safir::Time::TimeProvider::GetUtcTime());
        m_debug << boost::posix_time::to_simple_wstring(pt) << std::endl;

        if (m_connection.IsCreated(entityRequestProxy.GetEntityId()))
        {
            if (entityRequestProxy.GetTypeId() == Capabilities::Vehicles::Vehicle::ClassTypeId)
            {
                m_debug << "Sleeping for 1 seconds" << std::endl;
                ACE_OS::sleep(1);
            }
            else
            {
                m_debug << "Sleeping for 5 seconds" << std::endl;
                ACE_OS::sleep(5);
            }
            m_debug << "Done sleeping" << std::endl;

            m_connection.Delete(entityRequestProxy.GetEntityId(), 
                Safir::Dob::Typesystem::HandlerId());
        }
        responseSender -> Send(Safir::Dob::SuccessResponse::Create());       
    }
};

File: src/dose/dose_web.ss/src/com.saabgroup.safir.dob.web/src/com/saabgroup/safir/dob/web/Entity_Handler.java
// -*- coding: utf-8 -*-
/******************************************************************************
*
* Copyright Saab AB, 2007-2010 (http://www.safirsdk.com)
*
* Created by: Mikael Wennerberg / stmiwn
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
package com.saabgroup.safir.dob.web;

import com.saabgroup.safir.dob.Connection;
import com.saabgroup.safir.dob.Entity;
import com.saabgroup.safir.dob.EntityIterator;
import com.saabgroup.safir.dob.EntityProxy;
import com.saabgroup.safir.dob.Item;
import com.saabgroup.safir.dob.NotFoundException;
import com.saabgroup.safir.dob.Struct;
import com.saabgroup.safir.dob.typesystem.ContainerBase;
import com.saabgroup.safir.dob.typesystem.EntityId;
import com.saabgroup.safir.dob.typesystem.HandlerId;
import com.saabgroup.safir.dob.typesystem.InstanceId;
import com.saabgroup.safir.dob.typesystem.MemberInfo;
import com.saabgroup.safir.dob.typesystem.Members;
import com.saabgroup.safir.dob.typesystem.Operations;
import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import java.io.IOException;
import java.io.OutputStream;
import java.lang.reflect.InvocationTargetException;


class Entity_Handler implements HttpHandler {

    private Connection connection;

    public Entity_Handler() {
        connection = new Connection();
    }

    public void handle(HttpExchange exchange) throws IOException {
        // check the request method and process if it is a GET
        String requestMethod = exchange.getRequestMethod();
        if (requestMethod.equalsIgnoreCase("GET")) {
            // Set response headers
            Headers responseHeaders = exchange.getResponseHeaders();

            // Get response body
            OutputStream responseBody = exchange.getResponseBody();

            connection.open("webserver", "0", 0, null, null);
            responseHeaders.set("Content-Type", "text/html");
            //* response is OK (200)
            exchange.sendResponseHeaders(200, 0);

            responseBody.write("<html>".getBytes());
            responseBody.write("<body bgcolor=\"#EAEAEA\" lang=SV link=blue vlink=blue style='tab-interval:65.2pt'>".getBytes());
            responseBody.write(("<title>" + exchange.getRequestURI().getQuery() + "</title>").getBytes());

            if (exchange.getRequestURI().getPath().endsWith("show")) {


                responseBody.write(("<b>" + exchange.getRequestURI().getQuery() + "</b><br><br>").getBytes());

                long id = Operations.getTypeId(exchange.getRequestURI().getQuery());
                long count = connection.getNumberOfInstances(id, HandlerId.ALL_HANDLERS, false);

                if (count > 0) {
                    responseBody.write(("Count : " + count + "<br>").getBytes());
                    responseBody.write(("<b>Instances : </b><br>").getBytes());
                    EntityIterator iter = connection.getEntityIterator(id, false);
                    while (iter.hasNext()) {
                        EntityProxy prox = iter.next();
                        responseBody.write(("<A HREF=\"/dob/entity/view?" + prox.getEntityId().toString() + "\">" + prox.getEntityId().getInstanceId().toString() + "</A><br>").getBytes());
                        prox.dispose();
                    }
                    iter.dispose();
                } else {
                    responseBody.write(("No objects<br>").getBytes());
                }

            } else if (exchange.getRequestURI().getPath().endsWith("view")) {

                String[] query = exchange.getRequestURI().getQuery().split(",");
                InstanceId instanceId = new InstanceId(Long.parseLong(query[1].substring(0, query[1].length() - 1).trim()));
                long typeId = Operations.getTypeId(query[0].substring(1));
                EntityId id = new EntityId(typeId, instanceId);
                try {
                    EntityProxy prox = connection.read(id);

                    Entity entity = prox.getEntity();
                    responseBody.write(("<b>" + exchange.getRequestURI().getQuery() + "</b><br>").getBytes());
                    responseBody.write(GetObject(entity, false).getBytes());

                    prox.dispose();
                } catch (NotFoundException ex) {
                    responseBody.write("<b>Failed to read entity!</b>".getBytes());
                }
            } else {
                responseBody.write("<b>Not implemented!</b>".getBytes());
            }

            responseBody.write("</body></html>".getBytes());

            // Close the responseBody
            responseBody.close();
            connection.close();
        }
    }

    private final String GetObject(com.saabgroup.safir.dob.typesystem.Object obj, boolean useTab) {
        long typeId = obj.getTypeId();
        String response = "<br>";
        String tab = "";

        if (useTab) {
            tab = "&nbsp&nbsp";
        }

        for (int i = 0; i < Members.getNumberOfMembers(typeId); i++) {
            MemberInfo info = Members.getInfo(typeId, i);
            String memberName = Members.getName(typeId, i);
            ContainerBase cont;
            try {
                // array?
                if (info.getIsArray()) {
                    response = response.concat(("<b>" + memberName + " : </b><br>"));
                    for (int idx = 0; idx < info.getArrayLength(); idx++) {
                        cont = obj.getMember(i, idx);
                        response = response.concat(" &nbsp &nbsp ");
                        if (cont.isNull()) {
                            response = response.concat(("Index<b> " + idx + " :  NULL </b><br>"));
                        } else {
                            // item or struct?
                            if (Operations.isOfType(info.getMemberTypeId(), Item.ClassTypeId) || Operations.isOfType(info.getMemberTypeId(), Struct.ClassTypeId)) {
                                response = response.concat(("Index<b> " + idx + " : </b>" + GetObject((com.saabgroup.safir.dob.typesystem.Object) cont.getClass().getMethod("getObj", (Class<?>[]) null).invoke(cont, (Object[]) null), true)));
                            } else {
                                response = response.concat(("Index<b> " + idx + " : </b>" + cont.getClass().getMethod("getVal", (Class<?>[]) null).invoke(cont, (Object[]) null) + "<br>"));
                            }
                        }
                    }
                } else {
                    cont = obj.getMember(i, 0);
                    if (cont.isNull()) {
                        response = response.concat((tab + "<b>" + Members.getName(typeId, i) + " : NULL </b><br>"));
                    } else {
                        // item or struct?
                        if (Operations.isOfType(info.getMemberTypeId(), Item.ClassTypeId) || Operations.isOfType(info.getMemberTypeId(), Struct.ClassTypeId)) {
                            response = response.concat(("<b>" + Members.getName(typeId, i) + " : </b>" + GetObject((com.saabgroup.safir.dob.typesystem.Object) cont.getClass().getMethod("getObj", (Class<?>[]) null).invoke(cont, (Object[]) null), true)));
                        } else {
                            response = response.concat(tab + "<b>" + memberName + " : </b>" + cont.getClass().getMethod("getVal", (Class<?>[]) null).invoke(cont, (Object[]) null) + "<br>");
                        }
                    }
                }
            } catch (IllegalAccessException ex) {
                response = response.concat("<b>IllegalAccessException!</b><br>");
            } catch (IllegalArgumentException ex) {
                response = response.concat("<b>IllegalArgumentException!</b><br>");
            } catch (InvocationTargetException ex) {
                response = response.concat("<b>InvocationTargetException!</b><br>");
            } catch (NoSuchMethodException ex) {
                response = response.concat("<b>NoSuchMethodException!</b><br>");
            } catch (SecurityException ex) {
                response = response.concat("<b>SecurityException!</b><br>");
            }
        }

        return response;
    }
}
